crates/sp-lightclient/src/lib.rs

```diff
@@ -205,96 +229,274 @@ pub trait HeaderImporter<Header: HeaderT, Store: Storage<Header>> {
-            ErrorDigestType::Seal,
-        )))?;
-
-    let signature = seal
-        .as_subspace_seal()
-        .ok_or(ImportError::InvalidDigest(ErrorDigestType::Seal))?;
-
-    // The pre-hash of the header doesn't include the seal and that's what we sign
-    let pre_hash = header.hash();
-
-    // Verify that block is signed properly
-    check_reward_signature(
-        pre_hash.as_ref(),
-        &Into::<RewardSignature>::into(&signature),
-        &Into::<PublicKey>::into(public_key),
-        &schnorrkel::context::signing_context(REWARD_SIGNING_CONTEXT),
-    )
-    .map_err(|_| ImportError::InvalidBlockSignature)?;
-
-    // push the seal back into the header
-    header.digest_mut().push(seal);
-    Ok(())
-}
+            Some(number) => number,
+        };
 
-fn calculate_block_weight(
-    global_randomness: &Randomness,
-    pre_digest: &PreDigest<FarmerPublicKey, FarmerPublicKey>,
-) -> BlockWeight {
-    let global_challenge = derive_global_challenge(global_randomness, pre_digest.slot.into());
-
-    let target = u64::from_be_bytes(
-        derive_target(
-            &schnorrkel::PublicKey::from_bytes(pre_digest.solution.public_key.as_ref())
-                .expect("Always correct length; qed"),
-            global_challenge,
-            &pre_digest.solution.local_challenge,
-        )
-        .expect("Verification of the local challenge was done before this; qed"),
-    );
-    let tag = u64::from_be_bytes(pre_digest.solution.tag);
-    u128::from(u64::MAX - subspace_core_primitives::bidirectional_distance(&target, &tag))
+        match number_to_finalize.cmp(current_finalized_header.header.number()) {
+            Ordering::Less => Err(ImportError::SwitchedToForkBelowArchivingDepth),
+            // nothing to do as we finalized the header already
+            Ordering::Equal => Ok(()),
+            // finalize heads one after the other and prune any forks
+            Ordering::Greater => {
+                let mut current_finalized_number = *current_finalized_header.header.number();
+
+                while current_finalized_number < number_to_finalize {
+                    let number_to_finalize = current_finalized_number
+                        .checked_add(&One::one())
+                        .ok_or(ImportError::ArithmeticError(ArithmeticError::Overflow))?;
+
+                    // find the headers at the number to be finalized
+                    let headers_at_number_to_be_finalized =
+                        store.headers_at_number(number_to_finalize);
+                    // if there is just one header at that number, we mark that header as finalized and move one
+                    if headers_at_number_to_be_finalized.len() == 1 {
+                        let header_to_finalize = headers_at_number_to_be_finalized
+                            .into_iter()
+                            .next()
+                            .expect("Safe to unwrap because of the check above.");
```

Comment:
This can be simpler:

```suggestion
                            .first()
                            .expect("First item must exist as the len is 1.");
```